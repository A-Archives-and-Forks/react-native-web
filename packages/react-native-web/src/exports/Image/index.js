/**
 * Copyright (c) Nicolas Gallagher.
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import type { ImageProps, ImageStatics } from './types';
import type { PlatformMethods } from '../../types';

import * as React from 'react';
import createElement from '../createElement';
import css from '../StyleSheet/css';
import * as forwardedProps from '../../modules/forwardedProps';
import pick from '../../modules/pick';
import processColor from '../processColor';
import resolveShadowValue from '../StyleSheet/resolveShadowValue';
import useElementLayout from '../../modules/useElementLayout';
import useMergeRefs from '../../modules/useMergeRefs';
import usePlatformMethods from '../../modules/usePlatformMethods';
import useResponderEvents from '../../modules/useResponderEvents';
import ImageLoader from '../../modules/ImageLoader';
import PixelRatio from '../PixelRatio';
import StyleSheet from '../StyleSheet';
import TextAncestorContext from '../Text/TextAncestorContext';
import View from '../View';

import { getAssetByID, getAssetUriByID } from '../../modules/AssetRegistry';

const emptyObject = {};
const forwardPropsList = {
  ...forwardedProps.defaultProps,
  ...forwardedProps.accessibilityProps,
  ...forwardedProps.clickProps,
  ...forwardedProps.focusProps,
  ...forwardedProps.keyboardProps,
  ...forwardedProps.mouseProps,
  ...forwardedProps.touchProps,
  ...forwardedProps.styleProps,
  crossOrigin: true,
  decoding: true,
  draggable: true,
  loading: true,
  referrerPolicy: true
};

const pickProps = (props) => pick(props, forwardPropsList);

const svgDataUriPattern = /^(data:image\/svg\+xml;utf8,)(.*)/;

function getFlatStyle(style, blurRadius, resizeModeProp) {
  const initialStyle = StyleSheet.flatten(style) || emptyObject;
  const objectFitStyle = resizeModeStyles[initialStyle.resizeMode || resizeModeProp || 'cover'];
  const flatStyle = { ...initialStyle, ...objectFitStyle };
  const { filter, resizeMode, shadowOffset, tintColor } = flatStyle;

  // Add CSS filters
  // React Native exposes these features as props and proprietary styles
  const filters = [];
  if (filter) {
    filters.push(filter);
  }
  if (blurRadius) {
    filters.push(`blur(${blurRadius}px)`);
  }
  if (shadowOffset) {
    const shadowString = resolveShadowValue(flatStyle);
    if (shadowString) {
      filters.push(`drop-shadow(${shadowString})`);
    }
  }
  if (tintColor) {
    const colorInt = processColor(tintColor);
    if (colorInt != null) {
      const r = (colorInt >> 16) & 255;
      const g = (colorInt >> 8) & 255;
      const b = colorInt & 255;
      const a = ((colorInt >> 24) & 255) / 255;
      const alpha = a.toFixed(2);
      const matrix = `0 0 0 0 ${r / 255} 0 0 0 0 ${g / 255} 0 0 0 0 ${b / 255} 0 0 0 ${alpha} 0`;
      // NOTE: Safari doesn't support inline SVG filters (reported 2012)
      // https://bugs.webkit.org/show_bug.cgi?id=104169
      const svgFilter =
        "url('data:image/svg+xml," +
        '<svg xmlns="http://www.w3.org/2000/svg">' +
        '<filter id="tint">' +
        `<feColorMatrix type="matrix" values="${matrix}" />` +
        '</filter>' +
        "</svg>#tint')";
      filters.push(svgFilter);
    }
  }

  // These styles are converted to CSS filters applied to the
  // element displaying the background image.
  delete flatStyle.blurRadius;
  delete flatStyle.shadowColor;
  delete flatStyle.shadowOpacity;
  delete flatStyle.shadowOffset;
  delete flatStyle.shadowRadius;
  delete flatStyle.tintColor;
  // These styles are not supported on View
  delete flatStyle.overlayColor;
  delete flatStyle.resizeMode;

  if (filters.length > 0) {
    flatStyle.filter = filters.join(' ');
  }

  return flatStyle;
}

function getImageData(image: HTMLImageElement) {
  const { width, height, currentSrc } = image;
  return {
    source: { height, width, url: currentSrc },
    target: image
  }
}

function resolveAssetUri(source): ?string {
  let uri = null;
  if (typeof source === 'number') {
    uri = getAssetUriByID(source);
  } else if (typeof source === 'string') {
    uri = source;
  } else if (Array.isArray(source)) {
    uri = source[0].uri;
  } else if (source && typeof source.uri === 'string') {
    uri = source.uri;
  }
  if (uri) {
    const match = uri.match(svgDataUriPattern);
    // inline SVG markup may contain characters (e.g., #, ") that need to be escaped
    if (match) {
      const [, prefix, svg] = match;
      const encodedSvg = encodeURIComponent(svg);
      return `${prefix}${encodedSvg}`;
    }
  }
  return uri;
}

const ERROR_PLACEHOLDER =
  'data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==';

const Image: React.AbstractComponent<ImageProps, HTMLImageElement & PlatformMethods> = React.forwardRef(
  (props, forwardedRef) => {
    const {
      blurRadius,
      defaultSource,
      onError,
      onLayout,
      onLoad,
      onLoadEnd,
      onLoadStart,
      onMoveShouldSetResponder,
      onMoveShouldSetResponderCapture,
      onResponderEnd,
      onResponderGrant,
      onResponderMove,
      onResponderReject,
      onResponderRelease,
      onResponderStart,
      onResponderTerminate,
      onResponderTerminationRequest,
      onScrollShouldSetResponder,
      onScrollShouldSetResponderCapture,
      onSelectionChangeShouldSetResponder,
      onSelectionChangeShouldSetResponderCapture,
      onStartShouldSetResponder,
      onStartShouldSetResponderCapture,
      pointerEvents,
      resizeMode,
      source,
      style
    } = props;

    if (process.env.NODE_ENV !== 'production') {
      if (props.children) {
        throw new Error(
          'The <Image> component cannot contain children. If you want to render content on top of the image, consider using the <ImageBackground> component or absolute positioning.'
        );
      }
    }

    const supportedProps = pickProps(props);
    
    const src = resolveAssetUri(source) || resolveAssetUri(defaultSource);
    let srcSet;
    if (Array.isArray(source)) {
      srcSet = source.map(({ uri, scale }) => `${uri} ${scale || 1}x`);
    }

    const [managedSrc, setManagedSrc] = React.useState(src);
    const flatStyle = getFlatStyle(style, blurRadius, resizeMode);

    const hostRef = React.useRef(null);
    const hasTextAncestor = React.useContext(TextAncestorContext);
    useElementLayout(hostRef, onLayout);
    useResponderEvents(hostRef, {
      onMoveShouldSetResponder,
      onMoveShouldSetResponderCapture,
      onResponderEnd,
      onResponderGrant,
      onResponderMove,
      onResponderReject,
      onResponderRelease,
      onResponderStart,
      onResponderTerminate,
      onResponderTerminationRequest,
      onScrollShouldSetResponder,
      onScrollShouldSetResponderCapture,
      onSelectionChangeShouldSetResponder,
      onSelectionChangeShouldSetResponderCapture,
      onStartShouldSetResponder,
      onStartShouldSetResponderCapture
    });

    const internalImageRef = React.useCallback((target) => {
      const errorListener = function (e) {
        // If the image fails to load, browsers will display a "broken" icon.
        // To avoid this we replace the image with a transparent gif.
        setManagedSrc(ERROR_PLACEHOLDER);
        if (onError != null) {
          onError({
            nativeEvent: {
              error: `Failed to load resource ${e.target.src} (404)`
            }
          });
        }
        if (onLoadEnd != null) {
          onLoadEnd({ nativeEvent: { target }});
        }
      };

      const loadListener = function (e) {
        const { target: image } = e;
        if (image.src === ERROR_PLACEHOLDER) {
          // Prevent the placeholder from triggering a 'load' event that event
          // listeners would otherwise receive.
          e.stopImmediatePropagation();
        } else {
          if (onLoad != null) {
            onLoad({
              nativeEvent: getImageData(image)
            });
          }
          if (onLoadEnd != null) {
            onLoadEnd({ nativeEvent: { target }});
          }
        }
      };

      if (target !== null) {
        // If the image is loaded before JS loads (e.g., SSR), then we manually
        // call onLoad
//        console.log(target.complete)
        if (onLoad != null && target.complete) {
          onLoad({
            nativeEvent: getImageData(target)
          });
          return;
        }

        hostRef.current = target;
        if (onLoadStart != null) {
          onLoadStart({ nativeEvent: { target }});
        }
        target.addEventListener('error', errorListener);
        target.addEventListener('load', loadListener);
      } else if (hostRef.current != null) {
        const node = hostRef.current;
        node.removeEventListener('error', errorListener);
        node.removeEventListener('load', loadListener);
        hostRef.current = null;
      }
    },
    [onError, onLoad, onLoadEnd]
  );

    const platformMethodsRef = usePlatformMethods(supportedProps);

    const ref = useMergeRefs(
      hostRef,
      internalImageRef,
      platformMethodsRef,
      forwardedRef
    );

    supportedProps.alt = props.alternativeText;
    supportedProps.classList = hasTextAncestor ? inlineClassList : defaultClassList;
    supportedProps.decoding = props.decoding || 'async';
    supportedProps.draggable = props.draggable || false;
    supportedProps.loading = props.loading || 'lazy';
    supportedProps.ref = ref;
    supportedProps.src = managedSrc;
    supportedProps.srcSet = 
        srcSet != null && managedSrc !== ERROR_PLACEHOLDER
          ? srcSet.join(',')
          : null;
    supportedProps.style = flatStyle;

    return createElement('img', supportedProps);
  }
);

Image.displayName = 'Image';

const classes = css.create({
  image: {
    backgroundColor: 'transparent',
    border: '0 solid black',
    boxSizing: 'border-box',
    display: 'flex',
    flexDirection: 'column',
    flexShrink: 0,
    margin: 0,
    minHeight: 0,
    minWidth: 0,
    objectFit: 'cover',
    padding: 0,
    position: 'relative',
    zIndex: 0
  },
  inlineImage: {
    display: 'inline-flex'
  }
});

const defaultClassList = [classes.image];
const inlineClassList = [classes.image, classes.inlineImage];

const resizeModeStyles = {
  center: {
    objectFit: 'scale-down'
  },
  contain: {
    objectFit: 'contain'
  },
  cover: {
    objectFit: 'cover'
  },
  none: {
    objectFit: 'none'
  },
  stretch: {
    objectFit: 'fill'
  }
};

// $FlowIgnore: This is the correct type, but casting makes it unhappy since the variables aren't defined yet
const ImageWithStatics = (Image: React.AbstractComponent<
  ImageProps,
  React.ElementRef<typeof View>
> &
  ImageStatics);

ImageWithStatics.getSize = function (uri, success, failure) {
  return ImageLoader.getSize(uri, success, failure);
};

ImageWithStatics.prefetch = function (uri) {
  return ImageLoader.prefetch(uri);
};

ImageWithStatics.queryCache = function (uris) {
  return ImageLoader.queryCache(uris);
};

export type { ImageProps };

export default ImageWithStatics;
