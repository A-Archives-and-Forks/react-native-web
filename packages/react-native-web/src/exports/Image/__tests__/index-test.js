/* eslint-env jasmine, jest */
/* eslint-disable react/jsx-no-bind */

import { act } from 'react-dom/test-utils';
import * as AssetRegistry from '../../../modules/AssetRegistry';
import Image from '../';
import { ImageUriCache } from '../../../modules/ImageLoader';
import PixelRatio from '../../PixelRatio';
import React from 'react';
import { render } from '@testing-library/react';
import { createEventTarget } from 'dom-event-testing-library';

const originalImage = window.Image;

describe('components/Image', () => {
  beforeEach(() => {
    ImageUriCache._entries = {};
    window.Image = jest.fn(() => ({}));
  });

  afterEach(() => {
    window.Image = originalImage;
  });

  test('prop "accessibilityLabel"', () => {
    const defaultSource = { uri: 'https://google.com/favicon.ico' };
    const { container } = render(
      <Image accessibilityLabel="accessibilityLabel" defaultSource={defaultSource} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('prop "alternativeText"', () => {
    test('set to empty string', () => {
      const { container } = render(<Image alternativeText="" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('set to value', () => {
      const { container } = render(
        <Image alternativeText="alternative text" />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  test('prop "blurRadius"', () => {
    const defaultSource = { uri: 'https://google.com/favicon.ico' };
    const { container } = render(<Image blurRadius={5} defaultSource={defaultSource} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('prop "crossOrigin"', () => {
    test('sets value', () => {
      const { container } = render(<Image crossOrigin="use-credentials" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "decoding"', () => {
    test('sets value', () => {
      const { container } = render(<Image decoding="async" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "defaultSource"', () => {
    test('sets image when value is an object', () => {
      const defaultSource = { uri: 'https://google.com/favicon.ico' };
      const { container } = render(<Image defaultSource={defaultSource} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('sets image when value is a string', () => {
      // emulate require-ed asset
      const defaultSource = 'https://google.com/favicon.ico';
      const { container } = render(<Image defaultSource={defaultSource} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('sets "height" and "width" styles if missing', () => {
      const defaultSource = {
        uri: 'https://google.com/favicon.ico',
        height: 10,
        width: 20
      };
      const { container } = render(<Image defaultSource={defaultSource} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('does not override "height" and "width" styles', () => {
      const defaultSource = {
        uri: 'https://google.com/favicon.ico',
        height: 10,
        width: 20
      };
      const { container } = render(
        <Image defaultSource={defaultSource} style={{ height: 20, width: 40 }} />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  test('prop "draggable"', () => {
    const defaultSource = { uri: 'https://google.com/favicon.ico' };
    const { container } = render(<Image defaultSource={defaultSource} draggable={true} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('prop "focusable"', () => {
    const { container } = render(<Image focusable={true} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('prop "loading"', () => {
    test('sets value', () => {
      const { container } = render(<Image loading="lazy" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  test('prop "nativeID"', () => {
    const { container } = render(<Image nativeID="nativeID" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('prop "onError"', () => {
    test('is called when image fails to load and replaces src with placeholder', () => {
      const onError = jest.fn();
      const ref = React.createRef();
      let container;
      act(() => {
        ({ container } = render(
          <Image onError={onError} ref={ref} source="" />
        ));
      });
      const image = createEventTarget(ref.current);
      act(() => {
        image.error();
      });
      expect(onError).toBeCalledTimes(1);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "onLoad"', () => {
    test('is called once when image loads', () => {
      const onLoad = jest.fn();
      const onLoadEnd = jest.fn();
      const onLoadStart = jest.fn();
      const ref = React.createRef();
      let container;
      act(() => {
        ({ container } = render(
          <Image
            onLoad={onLoad}
            onLoadEnd={onLoadEnd}
            onLoadStart={onLoadStart}
            ref={ref}
            source="https://google.com/favicon.ico"
          />
        ));
      });
      const image = createEventTarget(ref.current);
      act(() => {
        image.load({
          target: {
            width: 100,
            height: 100,
            naturalWidth: 200,
            naturalHeight: 200,
            currentSrc: 'https://google.com/favicon.ico',
            src: 'https://google.com/favicon.ico'
          }
        });
      });
      expect(onLoad).toBeCalledTimes(1);
      expect(onLoadEnd).toBeCalledTimes(1);
      expect(onLoadStart).toBeCalledTimes(1);
      expect(onLoad).toBeCalledWith(
        expect.objectContaining({
          nativeEvent: expect.objectContaining({
            source: expect.objectContaining({
              width: expect.any(Number),
              height: expect.any(Number),
              url: 'https://google.com/favicon.ico'
            })
          })
        })
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    test('is called on update if "uri" is different', () => {
      const onLoadStartStub = jest.fn();
      const onLoadStub = jest.fn();
      const onLoadEndStub = jest.fn();
      const ref = React.createRef();
      act(() => {
        render(
          <Image
            onLoad={onLoadStub}
            onLoadEnd={onLoadEndStub}
            onLoadStart={onLoadStartStub}
            ref={ref}
            source={'https://test.com/img.jpg'}
          />
        );
      });
      const image = createEventTarget(ref.current);
      act(() => {
        image.load({
          target: {
            width: 100,
            height: 100,
            naturalWidth: 200,
            naturalHeight: 200,
            currentSrc: 'https://test.com/img.jpg',
            src: 'https://test.com/img.jpg'
          }
        });
      });

      expect(onLoadStub.mock.calls.length).toBe(1);
      expect(onLoadEndStub.mock.calls.length).toBe(1);
      expect(onLoadStartStub.mock.calls.length).toBe(1);

      act(() => {
        render(
          <Image
            onLoad={onLoadStub}
            onLoadEnd={onLoadEndStub}
            onLoadStart={onLoadStartStub}
            ref={ref}
            source={'https://blah.com/img.png'}
          />
        );
      });
      act(() => {
        image.load({
          target: {
            width: 100,
            height: 100,
            naturalWidth: 200,
            naturalHeight: 200,
            currentSrc: 'https://blah.com/img.png',
            src: 'https://blah.com/img.png'
          }
        });
      });

      expect(onLoadStub.mock.calls.length).toBe(2);
      expect(onLoadEndStub.mock.calls.length).toBe(2);
      expect(onLoadStartStub.mock.calls.length).toBe(2);
    });

    test('is not called when placeholder src is used after error', () => {
      const onError = jest.fn();
      const onLoad = jest.fn();
      const ref = React.createRef();
      act(() => {
        render(
          <Image
            onError={onError}
            onLoad={onLoad}
            ref={ref}
            source="https://google.com"
          />
        );
      });
      const image = createEventTarget(ref.current);
      act(() => {
        // Results in placeholder being "loaded"
        image.error();
      });
      expect(onError).toBeCalledTimes(1);
      act(() => {
        // Emulate the native "load" event for the placeholder
        image.load();
      });
      expect(onLoad).toBeCalledTimes(0);
    });
  });

  describe('prop "resizeMode"', () => {
    ['contain', 'cover', 'none', 'repeat', 'stretch', undefined].forEach((resizeMode) => {
      test(`value "${resizeMode}"`, () => {
        const { container } = render(<Image resizeMode={resizeMode} />);
        expect(container.firstChild).toMatchSnapshot();
      });
    });
  });

  describe('prop "source"', () => {
    test('does not throw', () => {
      const sources = [null, '', {}, { uri: '' }, { uri: 'https://google.com' }];
      sources.forEach((source) => {
        expect(() => render(<Image source={source} />)).not.toThrow();
      });
    });

    test('is not set immediately if the image has not already been loaded', () => {
      const uri = 'https://google.com/favicon.ico';
      const source = { uri };
      const { container } = render(<Image source={source} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('is set immediately if the image was preloaded', () => {
      const uri = 'https://yahoo.com/favicon.ico';
      //ImageLoader.load = jest.fn().mockImplementationOnce((_, onLoad, onError) => {
       // onLoad();
      //});
      return Image.prefetch(uri).then(() => {
        const source = { uri };
        const { container } = render(<Image source={source} />, { disableLifecycleMethods: true });
        expect(container.firstChild).toMatchSnapshot();
        ImageUriCache.remove(uri);
      });
    });

    test('is set immediately if the image has already been loaded', () => {
      const uriOne = 'https://google.com/favicon.ico';
      const uriTwo = 'https://twitter.com/favicon.ico';
      ImageUriCache.add(uriOne);
      ImageUriCache.add(uriTwo);

      // initial render
      const { container, rerender } = render(<Image source={{ uri: uriOne }} />);
      ImageUriCache.remove(uriOne);
      expect(container.firstChild).toMatchSnapshot();
      // props update
      act(() => {
        rerender(<Image source={{ uri: uriTwo }} />);
        ImageUriCache.remove(uriTwo);
      });
      expect(container.firstChild).toMatchSnapshot();
    });

    test('is correctly updated when missing in initial render', () => {
      const uri = 'https://testing.com/img.jpg';
      const { container, rerender } = render(<Image />);
      act(() => {
        rerender(<Image source={{ uri }} />);
      });
      expect(container.firstChild).toMatchSnapshot();
    });

    test('is correctly updated only when loaded if defaultSource provided', () => {
      const defaultUri = 'https://testing.com/preview.jpg';
      const uri = 'https://testing.com/fullSize.jpg';
      const { container } = render(<Image defaultSource={{ uri: defaultUri }} source={{ uri }} />);
      expect(container.firstChild).toMatchSnapshot();
      act(() => {
      //  loadCallback();
      });
      expect(container.firstChild).toMatchSnapshot();
    });

    test('it correctly selects the source scale', () => {
      AssetRegistry.getAssetByID = () => ({
        httpServerLocation: 'static',
        name: 'img',
        scales: [1, 2, 3],
        type: 'png'
      });

      PixelRatio.get = jest.fn(() => 1.0);
      let { container } = render(<Image source={1} />);
      expect(container.querySelector('img').src).toBe('http://localhost/static/img.png');

      act(() => {
        PixelRatio.get = jest.fn(() => 2.2);
        ({ container } = render(<Image source={1} />));
      });
      expect(container.querySelector('img').src).toBe('http://localhost/static/img@2x.png');
    });
  });

  describe('prop "style"', () => {
    test('supports "resizeMode" property', () => {
      const { container } = render(<Image style={{ resizeMode: 'contain' }} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('supports "shadow" properties (convert to filter)', () => {
      const { container } = render(
        <Image style={{ shadowColor: 'red', shadowOffset: { width: 1, height: 1 } }} />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    test('supports "tintcolor" property (convert to filter)', () => {
      const defaultSource = { uri: 'https://google.com/favicon.ico' };
      const { container } = render(
        <Image defaultSource={defaultSource} style={{ tintColor: 'red' }} />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    test('removes other unsupported View styles', () => {
      const { container } = render(<Image style={{ overlayColor: 'red', tintColor: 'blue' }} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  test('prop "testID"', () => {
    const { container } = render(<Image testID="testID" />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('queryCache', () => {
    const uriOne = 'https://google.com/favicon.ico';
    const uriTwo = 'https://twitter.com/favicon.ico';
    ImageUriCache.add(uriOne);
    ImageUriCache.add(uriTwo);
    return Image.queryCache([uriOne, uriTwo, 'oops']).then((res) => {
      expect(res).toEqual({
        [uriOne]: 'disk/memory',
        [uriTwo]: 'disk/memory'
      });
    });
  });
});
