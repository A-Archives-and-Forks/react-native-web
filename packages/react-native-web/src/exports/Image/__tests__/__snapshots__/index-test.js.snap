// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Image prop "accessibilityLabel" 1`] = `
<img
  aria-label="accessibilityLabel"
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "alternativeText" set to empty string 1`] = `
<img
  alt=""
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "alternativeText" set to value 1`] = `
<img
  alt="alternative text"
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "blurRadius" 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="filter: blur(5px); object-fit: cover;"
/>
`;

exports[`components/Image prop "crossOrigin" sets value 1`] = `
<img
  class="css-image-tcr4mf"
  crossorigin="use-credentials"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "decoding" sets value 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "defaultSource" does not override "height" and "width" styles 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="height: 20px; object-fit: cover; width: 40px;"
/>
`;

exports[`components/Image prop "defaultSource" sets "height" and "width" styles if missing 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "defaultSource" sets image when value is a string 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "defaultSource" sets image when value is an object 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "draggable" 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="true"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "focusable" 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: cover;"
  tabindex="0"
/>
`;

exports[`components/Image prop "loading" sets value 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "nativeID" 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  id="nativeID"
  loading="lazy"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "onError" is called when image fails to load and replaces src with placeholder 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=="
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "onLoad" is called once when image loads 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "resizeMode" value "contain" 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: contain;"
/>
`;

exports[`components/Image prop "resizeMode" value "cover" 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "resizeMode" value "none" 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: none;"
/>
`;

exports[`components/Image prop "resizeMode" value "repeat" 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
/>
`;

exports[`components/Image prop "resizeMode" value "stretch" 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: fill;"
/>
`;

exports[`components/Image prop "resizeMode" value "undefined" 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "source" is correctly updated only when loaded if defaultSource provided 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://testing.com/fullSize.jpg"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "source" is correctly updated only when loaded if defaultSource provided 2`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://testing.com/fullSize.jpg"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "source" is correctly updated when missing in initial render 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "source" is not set immediately if the image has not already been loaded 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "source" is set immediately if the image has already been loaded 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "source" is set immediately if the image has already been loaded 2`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="object-fit: cover;"
/>
`;

exports[`components/Image prop "source" is set immediately if the image was preloaded 1`] = `
<div
  class="css-view-1dbjc4n r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010"
>
  <div
    class="css-view-1dbjc4n r-backgroundColor-1niwhzg r-backgroundPosition-vvn4in r-backgroundRepeat-u6sd8q r-backgroundSize-4gszlv r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu r-zIndex-1wyyakw"
    style="background-image: url(https://yahoo.com/favicon.ico);"
  />
  <img
    alt=""
    class="css-accessibilityImage-9pa8cd"
    draggable="false"
    src="https://yahoo.com/favicon.ico"
  />
</div>
`;

exports[`components/Image prop "style" removes other unsupported View styles 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="filter: url('data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\"><filter id=\\"tint\\"><feColorMatrix type=\\"matrix\\" values=\\"0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1.00 0\\" /></filter></svg>#tint'); object-fit: cover;"
/>
`;

exports[`components/Image prop "style" supports "resizeMode" property 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: contain;"
/>
`;

exports[`components/Image prop "style" supports "shadow" properties (convert to filter) 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="filter: drop-shadow(1px 1px 0px rgba(255,0,0,1.00)); object-fit: cover;"
/>
`;

exports[`components/Image prop "style" supports "tintcolor" property (convert to filter) 1`] = `
<img
  class="css-image-tcr4mf"
  decoding="async"
  draggable="false"
  loading="lazy"
  src="https://google.com/favicon.ico"
  style="filter: url('data:image/svg+xml,<svg xmlns=\\"http://www.w3.org/2000/svg\\"><filter id=\\"tint\\"><feColorMatrix type=\\"matrix\\" values=\\"0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1.00 0\\" /></filter></svg>#tint'); object-fit: cover;"
/>
`;

exports[`components/Image prop "testID" 1`] = `
<img
  class="css-image-tcr4mf"
  data-testid="testID"
  decoding="async"
  draggable="false"
  loading="lazy"
  style="object-fit: cover;"
/>
`;
